File: test/objectives.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { evaluatePlayerObjectives } from '../cloudflare/src/objectives.ts';

test('Player Strategic Objectives Engine', async (t) => {
  await t.test('evaluates all 7 required long-term player objectives', () => {
    const objectives = evaluatePlayerObjectives({
      human: { credits: 12000, standing: 85, legacy: 300, voting_weight: 12.5 },
      business: { id: 'b-1', valuation: 45000, profit: 500 },
      institutions: {
        city: { essential_services_index: 0.85, standing: 80 },
        corporation: { treasury: 25000, member_count: 50 },
      },
      governance: { voting_weight: 12.5 },
      technology: { active_patents: 2, active_licenses: 3 },
      dynasty: { generation: 2, perks_count: 2, successor_id: 'H-0099' },
      map: { plots_leased: 3 },
      resources: { material: 120 },
      netWorth: 28000,
    });

    assert.equal(objectives.length, 7);

    const ids = objectives.map((o) => o.id);
    assert.ok(ids.includes('obj-valuable-corporation'));
    assert.ok(ids.includes('obj-civic-delegate'));
    assert.ok(ids.includes('obj-regional-resource-control'));
    assert.ok(ids.includes('obj-dynasty-traits'));
    assert.ok(ids.includes('obj-technology-licensor'));
    assert.ok(ids.includes('obj-financial-independence'));
    assert.ok(ids.includes('obj-public-service-score'));

    for (const obj of objectives) {
      assert.ok(obj.id);
      assert.ok(obj.title);
      assert.ok(obj.description);
      assert.ok(typeof obj.currentValue === 'number');
      assert.ok(typeof obj.targetValue === 'number');
      assert.ok(typeof obj.progressPercentage === 'number');
      assert.ok(obj.progressPercentage >= 0 && obj.progressPercentage <= 100);
      assert.ok(obj.metricLabel);
      assert.ok(['in_progress', 'completed', 'claimed'].includes(obj.status));
      assert.ok(obj.rewardDescription);
      assert.ok(obj.targetSection);
      assert.ok(obj.iconName);
    }
  });

  await t.test('detects completed objectives when thresholds are reached', () => {
    const objectives = evaluatePlayerObjectives({
      human: { credits: 100000, standing: 98, voting_weight: 30 },
      business: { valuation: 150000, profit: 5000 },
      institutions: {
        city: { essential_services_index: 0.95, standing: 95 },
      },
      governance: { voting_weight: 30 },
      technology: { active_patents: 4, active_licenses: 4 },
      dynasty: { generation: 3, perks_count: 4, successor_id: 'H-0099' },
      map: { plots_leased: 6 },
      resources: { material: 500 },
      netWorth: 80000,
    });

    assert.equal(objectives.length, 7);
    assert.ok(objectives.some((o) => o.status === 'completed'));

    const corpObj = objectives.find((o) => o.id === 'obj-valuable-corporation');
    assert.equal(corpObj.status, 'completed');
    assert.equal(corpObj.progressPercentage, 100);

    const financeObj = objectives.find((o) => o.id === 'obj-financial-independence');
    assert.equal(financeObj.status, 'completed');
    assert.equal(financeObj.progressPercentage, 100);
  });
});

File: flutter_client/test/objectives_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:earth_client/core/models/player_objective.dart';
import 'package:earth_client/core/models/earth_state.dart';
import 'package:earth_client/features/command_center/objectives_panel.dart';

void main() {
  final testState = const EarthState({
    'clock': {'day': 185, 'minute': 400},
    'world': {'health': 80, 'essentialServicesIndex': 0.88},
    'human': {'id': 'H-0044', 'name': 'Amara Vance', 'credits': 32000.0, 'standing': 85.0},
    'resources': {'energy': 80.0, 'material': 150.0, 'food': 50.0},
    'business': {'id': 'b-1', 'name': 'AeroWorks', 'profit': 1500.0},
    'contracts': [
      {'id': 'c-101', 'title': 'Aerospace Composites', 'status': 'active'}
    ],
    'machines': [
      {'id': 'm-1', 'name': 'Precision Lathe', 'condition': 85.0, 'utilization': 80.0},
      {'id': 'm-2', 'name': 'Foundry Mk2', 'condition': 90.0, 'utilization': 70.0}
    ],
    'governance': {
      'proposals': [
        {'id': 'prop-1', 'title': 'Municipal Tax Charter Revision', 'status': 'open'}
      ]
    },
    'technology': {
      'research': {'progress': 75.0}
    },
    'life': {
      'generation': 2,
      'status': 'active',
      'successor': {'id': 'H-0099', 'name': 'Cyrus Vance'},
    },
  });

  group('PlayerObjective Model & Synthesis', () {
    test('serializes and deserializes JSON correctly', () {
      const obj = PlayerObjective(
        id: 'obj-1',
        category: 'enterprise',
        title: 'Build the Most Valuable Corporation',
        description: 'Grow your enterprise into an industrial conglomerate.',
        currentValue: 45000.0,
        targetValue: 100000.0,
        progressPercentage: 45.0,
        metricLabel: '45,000 / 100,000 C Valuation',
        status: 'in_progress',
        rewardDescription: 'Title: "Industrial Titan"',
        targetSection: 'business',
        iconName: 'business_center',
      );

      final json = obj.toJson();
      final fromJson = PlayerObjective.fromJson(json);

      expect(fromJson.id, 'obj-1');
      expect(fromJson.category, 'enterprise');
      expect(fromJson.title, 'Build the Most Valuable Corporation');
      expect(fromJson.progressPercentage, 45.0);
      expect(fromJson.categoryLabel, 'ENTERPRISE & INDUSTRY');
      expect(fromJson.categoryColor, const Color(0xFFF59E0B));
      expect(fromJson.categoryIcon, Icons.business_center_outlined);
      expect(fromJson.isCompleted, isFalse);
    });

    test('synthesizes all exemplar objectives from state', () {
      final objectives = PlayerObjective.synthesizeFromState(testState);

      expect(objectives.length, 9);

      final titles = objectives.map((o) => o.title).toList();
      expect(titles.any((t) => t.contains('Valuable Corporation')), isTrue);
      expect(titles.any((t) => t.contains('Civic Delegate')), isTrue);
      expect(titles.any((t) => t.contains('House with Sovereign Traits')), isTrue);
      expect(titles.any((t) => t.contains('Technology Licensor')), isTrue);
      expect(titles.any((t) => t.contains('Financial Independence')), isTrue);
      expect(titles.any((t) => t.contains('Public-Service Score')), isTrue);
    });
  });

  group('ObjectivesPanel Widget', () {
    testWidgets('renders ambition gauge, cards, reward descriptions, and triggers pursue action',
        (tester) async {
      final objectives = PlayerObjective.synthesizeFromState(testState);
      String? navigatedSection;

      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: SingleChildScrollView(
              child: ObjectivesPanel(
                objectives: objectives,
                onNavigate: (s) => navigatedSection = s,
              ),
            ),
          ),
        ),
      );

      // Verify Header and Global Ambition Gauge
      expect(find.text('CURRENT DIRECTION'), findsOneWidget);
      expect(find.textContaining('DIRECTIONS COMPLETED'), findsOneWidget);
      expect(find.textContaining('OVERALL PROGRESS'), findsOneWidget);

      // Verify Filter Pills
      expect(find.textContaining('ALL (9)'), findsOneWidget);
      expect(find.text('ENTERPRISE'), findsOneWidget);
      expect(find.text('CIVIC & HOUSE'), findsOneWidget);
      expect(find.text('TECH & FINANCE'), findsOneWidget);
      expect(find.textContaining('COMPLETED ('), findsOneWidget);

      // Verify Titles of Objectives
      expect(find.text('Build the Most Valuable Corporation'), findsOneWidget);
      expect(find.text('Become a Major Civic Delegate'), findsOneWidget);
      expect(find.text('Found a House with Sovereign Traits'), findsOneWidget);
      expect(find.text('Become a Leading Technology Licensor'), findsNothing);
      expect(find.text('Reach Financial Independence'), findsNothing);
      expect(find.text('Maintain the Highest Public-Service Score'), findsNothing);

      // Verify Rewards preview text
      expect(find.textContaining('Industrial Titan'), findsOneWidget);
      expect(find.textContaining('Grand Tribune'), findsOneWidget);

      // Test Pursue Button Click
      final pursueBtn = find.text('PURSUE').first;
      await tester.ensureVisible(pursueBtn);
      await tester.pumpAndSettle();
      await tester.tap(pursueBtn);
      await tester.pump();

      expect(navigatedSection, isNotNull);

      // Test Filtering Tabs
      final techFinanceTab = find.text('TECH & FINANCE');
      await tester.ensureVisible(techFinanceTab);
      await tester.pumpAndSettle();
      await tester.tap(techFinanceTab);
      await tester.pumpAndSettle();

      // Technology & Finance should be present
      expect(find.text('Become a Leading Technology Licensor'), findsOneWidget);
      expect(find.text('Reach Financial Independence'), findsOneWidget);
      // Civic delegate should be filtered out
      expect(find.text('Become a Major Civic Delegate'), findsNothing);
    });

    testWidgets('showObjectivesDialog opens modal and dismisses', (tester) async {
      final objectives = PlayerObjective.synthesizeFromState(testState);

      await tester.pumpWidget(
        MaterialApp(
          home: Builder(
            builder: (context) => Scaffold(
              body: Center(
                child: ElevatedButton(
                  onPressed: () => showObjectivesDialog(
                    context,
                    objectives: objectives,
                  ),
                  child: const Text('OPEN CODEX'),
                ),
              ),
            ),
          ),
        ),
      );

      await tester.tap(find.text('OPEN CODEX'));
      await tester.pumpAndSettle();

      expect(find.byType(Dialog), findsOneWidget);
      expect(find.text('LONG-TERM STRATEGIC OBJECTIVES'), findsWidgets);

      // Close modal
      await tester.tap(find.byIcon(Icons.close));
      await tester.pumpAndSettle();

      expect(find.byType(Dialog), findsNothing);
    });
  });
}
